osu.Game/Overlays/DirectOverlay.cs

```diff
@@ -244,54 +249,120 @@ protected override void PopIn()
-
         private readonly Bindable<string> currentQuery = new Bindable<string>(string.Empty);
-        private int currentPage = 1;
-        private bool isLastPageLoaded = false;
 
         private ScheduledDelegate queryChangedDebounce;
         private ScheduledDelegate addPageDebounce;
         private PreviewTrackManager previewTrackManager;
+        private BeatmapSetPager beatmapSetPager;
+
+        private class BeatmapSetPager
+        {
+            public event PageFetchHandler PageFetch;
+
+            private readonly RulesetStore rulesets;
+
+            private SearchBeatmapSetsRequest getSetsRequest;
+
+            private int currentPage = 1;
+
+            public bool IsLastPageFetched { get; private set; } = false;
+            public bool IsFetching => getSetsRequest != null;
+
+            public BeatmapSetPager(RulesetStore rulesets)
+            {
+                this.rulesets = rulesets;
+            }
+
+            public SearchBeatmapSetsRequest FetchNextPage(string query, RulesetInfo ruleset, BeatmapSearchCategory searchCategory = BeatmapSearchCategory.Any, DirectSortCriteria sortCriteria = DirectSortCriteria.Ranked, SortDirection direction = SortDirection.Descending)
+            {
+                if (getSetsRequest != null)
+                    return null;
+
+                getSetsRequest = new SearchBeatmapSetsRequest(
+                    query,
+                    ruleset,
+                    currentPage,
+                    searchCategory,
+                    sortCriteria,
+                    direction);
+
+                getSetsRequest.Success += response =>
+                {
+                    Task.Run(() =>
+                    {
+                        var sets = response.BeatmapSets.Select(r => r.ToBeatmapSet(rulesets)).ToList();
+
+                        if (sets.Count <= 0) IsLastPageFetched = true;
+
+                        PageFetch?.Invoke(currentPage, sets);
+                        
+                        getSetsRequest = null;
+                        currentPage++;
+                    });
+                };
+
+                return getSetsRequest;
+            }
+
+            public void Reset()
+            {
+                IsLastPageFetched = false;
+
+                currentPage = 1;
+
+                getSetsRequest?.Cancel();
+                getSetsRequest = null;
+            }
+
+            public delegate void PageFetchHandler(int page, List<BeatmapSetInfo> sets);
+        }
+
+        private void onPageFetch(int page, List<BeatmapSetInfo> sets)
+        {
+            if (page > 1)
+            {
+                BeatmapSets = BeatmapSets.Concat(sets);
+                addPanels(Filter.DisplayStyleControl.DisplayStyle.Value, sets);
+            }
+            else
+            {
+                BeatmapSets = sets;
+                recreatePanels(Filter.DisplayStyleControl.DisplayStyle.Value);
+            }
```

Comment:
Should this go on the scheduler thread?

Reply:
That would probably be correct, yes, to prevent mutating UI from a non-UI thread.